File: frontend/src/api/organigrama.ts
import { fetchWithAuth } from "./http.ts";

export type OrganigramaNodo = {
  id: number;
  empleado_id: number;
  no_empleado: string;
  nombre_colaborador: string;
  nombre_puesto: string | null;
  departamento: string | null;
  correo: string | null;
  foto_url: string | null;
  extension_telefono: string | null;
  parent_id: number | null;
  nivel_jerarquico: number;
  nivel_visual: "direccion" | "gerencia" | "jefaturas" | "operacion";
  activo: boolean;
  estado_empleado: string | null;
  reportes_directos: number;
  created_at: string;
  updated_at: string | null;
  relacion_incompleta: boolean;
  children: OrganigramaNodo[];
};

export type OrganigramaResponse = {
  total_nodos: number;
  total_raices: number;
  total_relaciones_incompletas: number;
  generated_at: string;
  roots: OrganigramaNodo[];
};

export type OrganigramaFetchError = {
  status: number;
  detail: string;
};

function throwOrganigramaError(status: number, detail: string): never {
  const err: OrganigramaFetchError = { status, detail };
  throw err;
}

async function readErrorDetail(res: Response): Promise<string> {
  const raw = await res.text();
  try {
    const parsed = JSON.parse(raw) as { detail?: unknown };
    if (typeof parsed.detail === "string") return parsed.detail;
  } catch {
    /* ignore */
  }
  return raw || res.statusText || "Error";
}

export async function getOrganigrama(): Promise<OrganigramaResponse> {
  const res = await fetchWithAuth("/api/v1/organigrama");
  if (!res.ok) {
    throwOrganigramaError(res.status, await readErrorDetail(res));
  }
  return (await res.json()) as OrganigramaResponse;
}

File: app/schemas/organigrama.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class OrganigramaNodoResponse(BaseModel):
    id: int
    empleado_id: int
    no_empleado: int
    nombre_colaborador: str
    nombre_puesto: str | None = None
    departamento: str | None = None
    correo: str | None = None
    foto_url: str | None = None
    extension_telefono: str | None = None
    parent_id: int | None = None
    nivel_jerarquico: int = 0
    nivel_visual: str = "operacion"
    activo: bool
    estado_empleado: str | None = None
    reportes_directos: int = 0
    created_at: datetime
    updated_at: datetime | None = None
    relacion_incompleta: bool = False
    children: list["OrganigramaNodoResponse"] = Field(default_factory=list)


class OrganigramaResponse(BaseModel):
    total_nodos: int
    total_raices: int
    total_relaciones_incompletas: int
    generated_at: datetime
    roots: list[OrganigramaNodoResponse]


OrganigramaNodoResponse.model_rebuild()
